packages/dds/tree/src/util/breakable.ts

```diff
@@ -0,0 +1,129 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { UsageError } from "@fluidframework/telemetry-utils/internal";
+
+/**
+ * An object which can enter a "broken" state where trying to use it is a UsageError.
+ */
+export class Breakable {
+	private brokenBy?: Error;
+
+	public constructor(private readonly name: string) {}
+
+	/**
+	 * Throws if the object is in the broken state.
+	 * @remarks
+	 * Can use {@link throwIfBroken} to apply this to a method.
+	 */
+	public use(): void {
+		if (this.brokenBy !== undefined) {
+			throw new UsageError(
+				`Invalid use of ${this.name} after it was put into an invalid state by another error.\nOriginal Error:\n${this.brokenBy}`,
+			);
+		}
+	}
+
+	/**
+	 * Puts this object into the broken state, and throws an error.
+	 *
+	 * @throws If already broken by a different error, throws a UsageError, otherwise throws `brokenBy`.
+	 */
+	public break(brokenBy: Error): never {
+		// If already broken by this error, let it bubble up without rethrowing a modified version.
+		// This prevents internal errors like asserts getting rethrow different errors when wrapped with multiple catches to ensure brokenBy is set.
```

Comment:
```suggestion
		// This prevents internal errors like asserts getting rethrown as different errors when wrapped with multiple call to `{@link Breakable.run}` or `{@link breakingMethod}`.
```
I found the "to ensure `brokenBy` is set" bit somewhat confusing since it was not clear, from a purely grammatical standpoint, whether it applied to "This prevents internal errors" or "catches". Having read the code, my understanding is that it's the latter. I don't think there's a strong need to be explicit about that here (the reader can go read the docs for `Breakable.run` or `breakingMethod`).